contracts/button/button.rs

```diff
@@ -0,0 +1,399 @@
+            .exec_input(
+                ExecutionInput::new(Selector::new(transfer_to_selector))
+                    .push_arg(to)
+                    .push_arg(value),
+            )
+            .returns::<()>()
+            .fire()
+    }
+
+    fn check_role(&self, account: AccountId, role: Role) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        <Self as AccessControlled>::check_role(
+            self.get().access_control,
+            account,
+            role,
+            |why: InkEnvError| {
+                Error::ContractCall(format!("Calling access control has failed: {:?}", why))
+            },
+            || Error::MissingRole,
+        )
+    }
+
+    fn allow(&mut self, player: AccountId, caller: AccountId, this: AccountId) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        let required_role = Role::Admin(this);
+        self.check_role(caller, required_role)?;
+        self.get_mut().can_play.insert(player, &true);
+        Ok(())
+    }
+
+    fn bulk_allow(
+        &mut self,
+        players: Vec<AccountId>,
+        caller: AccountId,
+        this: AccountId,
+    ) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        let required_role = Role::Admin(this);
+        self.check_role(caller, required_role)?;
+
+        for player in players {
+            self.get_mut().can_play.insert(player, &true);
+        }
+        Ok(())
+    }
+
+    fn disallow(&mut self, player: AccountId, caller: AccountId, this: AccountId) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        let required_role = Role::Admin(this);
+        self.check_role(caller, required_role)?;
+
+        self.get_mut().can_play.insert(player, &false);
+        Ok(())
+    }
+
+    fn press(&mut self, now: u32, caller: AccountId) -> Result<()> {
+        let ButtonData {
+            can_play, presses, ..
+        } = self.get();
+
+        if self.is_dead(now) {
+            return Err(Error::AfterDeadline);
+        }
+
+        if presses.get(&caller).is_some() {
+            return Err(Error::AlreadyParticipated);
+        }
+
+        if !can_play.get(&caller).unwrap_or(false) {
+            return Err(Error::NotWhitelisted);
+        }
+
+        let score = self.score(now);
+
+        self.get_mut().presses.insert(&caller, &score);
+        self.get_mut().press_accounts.push(caller);
+        self.get_mut().last_presser = Some(caller);
+        self.get_mut().last_press = now;
+        self.get_mut().total_scores += score;
+
+        Ok(())
+    }
+
+    fn death(
+        &self,
+        balance_of_selector: [u8; 4],
+        transfer_selector: [u8; 4],
+        this: AccountId,
+    ) -> Result<()> {
+        let ButtonData {
+            total_scores,
+            last_presser,
+            press_accounts,
+            presses,
+            ..
+        } = self.get();
+
+        let total_balance = self.balance(balance_of_selector, this)?;
```

Comment:
maybe we should check firstly whether there is any pressiah to avoid unnecessary call in case when nobody has pressed the button

Reply:
that would be sad, but we can do that optimization :)